core/jvm/src/main/scala/fs2/compression.scala

```diff
@@ -0,0 +1,863 @@
+package fs2
+
+import java.io.EOFException
+import java.nio.charset.StandardCharsets
+import java.time.Instant
+import java.util.zip._
+
+import cats.effect.Sync
+
+/** Provides utilities for compressing/decompressing byte streams. */
+object compression {
+
+  /**
+    * Returns a `Pipe` that deflates (compresses) its input elements using
+    * a `java.util.zip.Deflater` with the parameters `level`, `nowrap` and `strategy`.
+    * @param level the compression level (0-9)
+    * @param nowrap if true then use GZIP compatible compression
+    * @param bufferSize size of the internal buffer that is used by the
+    *                   compressor. Default size is 32 KB.
+    * @param strategy compression strategy -- see `java.util.zip.Deflater` for details
+    */
+  def deflate[F[_]](
+      level: Int = Deflater.DEFAULT_COMPRESSION,
+      nowrap: Boolean = false,
+      bufferSize: Int = 1024 * 32,
+      strategy: Int = Deflater.DEFAULT_STRATEGY
+  )(implicit SyncF: Sync[F]): Pipe[F, Byte, Byte] =
+    stream =>
+      Stream
+        .bracket(
+          SyncF.delay {
+            val deflater = new Deflater(level, nowrap)
+            deflater.setStrategy(strategy)
+            deflater
+          }
+        )(deflater => SyncF.delay(deflater.end()))
+        .flatMap(deflater => _deflate(deflater, bufferSize, crc32 = None)(stream))
+
+  private def _deflate[F[_]](
+      deflater: Deflater,
+      bufferSize: Int,
+      crc32: Option[CRC32]
+  ): Pipe[F, Byte, Byte] =
+    _deflate_stream(deflater, bufferSize.max(minBufferSize), crc32)(_).stream
+
+  private def _deflate_chunk[F[_]](
+      chunk: Chunk[Byte],
+      deflater: Deflater,
+      deflatedBufferSize: Int,
+      isFinalChunk: Boolean,
+      crc32: Option[CRC32]
+  ): Pull[F, Byte, Unit] = {
+    val bytesChunk = chunk.toBytes
+    deflater.setInput(
+      bytesChunk.values,
+      bytesChunk.offset,
+      bytesChunk.length
+    )
+    if (isFinalChunk)
+      deflater.finish()
+    crc32.foreach(_.update(bytesChunk.values, bytesChunk.offset, bytesChunk.length))
+
+    def isDone: Boolean =
+      (isFinalChunk && deflater.finished) || (!isFinalChunk && deflater.needsInput)
+
+    def deflateInto(deflatedBuffer: Array[Byte]): Int =
+      if (isDone) 0 else deflater.deflate(deflatedBuffer)
+
+    def pull(): Pull[F, Byte, Unit] = {
+      val deflatedBuffer = new Array[Byte](deflatedBufferSize)
```

Comment:
I saw this in the previous implementation as well, isn't this a bottleneck?

If I stream `n` GiB of data, this is going to allocate and garbage collect `n` GiB of memory. What are your thoughts on using a single mutable `Array` as a buffer, instead of allocating and garbage collecting a new buffer on each pull?

Reply:
It defaults to 32 KB in size. It also gets wrapped up and emitted as the output chunk, so it cannot be shared.

Reply:
BTW, I implemented the shared buffer approach before merging and then realized exactly what Daniel pointed out -- the buffer is emitted to the output so if we use a shared buffer, then we'd need to copy it when creating a chunk.